File: Samples/C#/BeamParams/BeamParamsPlugin.cs
using System;
using System.Text;

namespace BeamParams
{
  public class BeamParamsPlugin : Renga.IPlugin
  {
    private Renga.Application m_app;
    private Renga.ActionEventSource m_events;

    public bool Initialize(string pluginFolder)
    {
      m_app = new Renga.Application();
      var ui = m_app.UI;
      var panelExtension = ui.CreateUIPanelExtension();
      var action = ui.CreateAction();
      action.ToolTip = "Beam Params";
      m_events = new Renga.ActionEventSource(action);
      m_events.Triggered += (s, e) =>
      {
        ui.ShowMessageBox(Renga.MessageIcon.MessageIcon_Info,
                          "Beam Params",
                          buildBeamParamsListDescription());
      };
      panelExtension.AddToolButton(action);
      ui.AddExtensionToPrimaryPanel(panelExtension);
      return true;
    }

    public void Stop()
    {
      m_events.Dispose();
    }

    private Renga.Vector3D GetBaselineOffset(Renga.IBeamParams beamParams, Renga.Point2D offset_)
    {
      var offset = beamParams.GetProfilePlacement().GetTransformFrom().TransformPoint(offset_);

      var baseLine = beamParams.GetBaseline();

      var sectionPlacement = beamParams.GetProfilePlacementOnBaseline(baseLine.MinParameter);

      var sectionPlacementTransform = sectionPlacement.GetTransformFrom();

      var offsetPoint = sectionPlacementTransform.TransformPoint(new Renga.Point3D { X = offset.X, Y = offset.Y, Z = 0 });
      var baselineStart = baseLine.GetBeginPoint();

      return new Renga.Vector3D
      {
        X = offsetPoint.X - baselineStart.X,
        Y = offsetPoint.Y - baselineStart.Y,
        Z = offsetPoint.Z - baselineStart.Z,
      };
    }

    private string buildBeamParamsDescription(Renga.IBeamParams beamParams, Renga.IBeamStyleManager beamStyleManager, Renga.IPlacement3D levelPlacement)
    {
      var result_output = new StringBuilder("  Beam Param:\n");

      var transformIntoGlobal = levelPlacement.GetTransformFrom();
      var transformIntoGlobalWithVerticalOffset = transformIntoGlobal.GetCopy();
      transformIntoGlobalWithVerticalOffset.AddTranslation(new Renga.Vector3D { X = 0, Y = 0, Z = beamParams.VerticalOffset });

      // getting the baseline of the beam
      var baseLineInLevelSpace = beamParams.GetBaseline();

      // getting beam profile
      var profile = beamStyleManager.GetBeamStyle(beamParams.StyleId).Profile;
      var profilePlacement = beamParams.GetProfilePlacement();

      // curves in the global coordinate system

      // getting the baseline of the beam in the global coordinate system
      var baseLine = baseLineInLevelSpace.GetTransformed(transformIntoGlobal);

      // getting the center line of the beam in the global coordinate system
      var centerLine = baseLineInLevelSpace.GetOffseted(GetBaselineOffset(beamParams, new Renga.Point2D { X = 0, Y = 0 })).GetTransformed(transformIntoGlobalWithVerticalOffset);

      // getting the center line of mass of the beam in the global coordinate system
      var centerOfMassPath = baseLineInLevelSpace.GetOffseted(GetBaselineOffset(beamParams, profile.GetCenterOfMass())).GetTransformed(transformIntoGlobalWithVerticalOffset);

      // getting the length of the beam
      result_output.AppendLine("  beam length = " + centerOfMassPath.GetLength());
      result_output.AppendLine("");

      // getting the parameter of the corresponding beam start
      double parameter = baseLineInLevelSpace.MinParameter;

      var profilePlacementOnBaseline = beamParams.GetProfilePlacementOnBaseline(parameter);
      for (int i = 0; i < profile.Regions.Count; ++i)
      {
        var currentRegion = profile.Regions.Get(i);
        for (int j = 0; j < currentRegion.GetContourCount(); ++j)
        {
          var currentContour = currentRegion.GetContour(j).GetTransformed(profilePlacement.GetTransformFrom());

          // obtaining the curve of one of the regions of the beam profile at the beginning of the beam
          var sectionCurve = currentContour.CreateCurve3D(profilePlacementOnBaseline).GetTransformed(transformIntoGlobalWithVerticalOffset);
        }
      }

      return result_output.ToString();
    }

    private Renga.IPlacement3D getLevelPlacement(Renga.IModelObjectCollection modelObjects, Renga.IModelObject modelObject)
    {
      var parentLevelId = (modelObject as Renga.ILevelObject).LevelId;
      var parentLevel = modelObjects.GetById(parentLevelId);
      return (parentLevel as Renga.ILevel).Placement;
    }

    private string buildBeamParamsListDescription()
    {
      var project = m_app.Project;

      if (project == null)
        return "The Renga project is null";

      var result_output = new StringBuilder("Beam Params:\n");

      var beamStyleManager = project.BeamStyleManager;

      var modelObjects = project.Model.GetObjects();
      for (int i = 0; i < modelObjects.Count; i++)
      {
        var modelObject = modelObjects.GetByIndex(i);

        var beamParams = modelObject as Renga.IBeamParams;
        if (beamParams != null)
        {
          var levelPlacement = getLevelPlacement(modelObjects, modelObject);
          result_output.AppendLine(buildBeamParamsDescription(beamParams, beamStyleManager, levelPlacement));
        }
      }

      return result_output.ToString();
    }
  }
}

File: Samples/C++/BeamParams/BeamParamsPlugin.cpp
/*  Sample: Object properties
 *
 *  This sample illustrates how to work with beam params via Renga API.
 *
 *  Copyright Renga Software LLC, 2018. All rights reserved.
 */

#include "BeamParamsPlugin.h"

#include <sstream>

#include <Renga/ObjectTypes.h>
#include <Renga/CreateApplication.hpp>


BeamParamsPlugin::BeamParamsPlugin()
{
  ::CoInitializeEx(nullptr, COINIT_APARTMENTTHREADED);
}

BeamParamsPlugin::~BeamParamsPlugin()
{
  ::CoUninitialize();
}

bool BeamParamsPlugin::initialize(const wchar_t* pluginDir)
{
  auto pApplication = Renga::CreateApplication();
  if (!pApplication)
    return false;

  if (auto pUI = pApplication->GetUI())
  {
    if (auto pUIPanelExtension = pUI->CreateUIPanelExtension())
    {
      if (auto pAction = pUI->CreateAction())
      {
        pAction->PutToolTip(L"Beam Params.");

        m_pButtonClickHandler.reset(new BeamParamsClickHandler(pAction, pApplication));

        pUIPanelExtension->AddToolButton(pAction);
      }
      pUI->AddExtensionToPrimaryPanel(pUIPanelExtension);
    }
  }
  return true;
}


void BeamParamsPlugin::stop()
{
  m_pButtonClickHandler.reset();
}


BeamParamsClickHandler::BeamParamsClickHandler(Renga::IActionPtr pAction, Renga::IApplicationPtr pApplication) :
  Renga::ActionEventHandler(pAction),
  m_pApplication(pApplication)
{
}


void BeamParamsClickHandler::OnTriggered()
{
  const auto command_output = addPropertyToLevelObjects();

  MessageBox(NULL,
    (LPCWSTR)command_output.c_str(),
    (LPCWSTR)L"Beam Params",
    MB_ICONINFORMATION | MB_OK );
}

Renga::Vector3D BeamParamsClickHandler::GetBaselineOffset(Renga::IBeamParamsPtr pBeamParams, Renga::Point2D offset_)
{
  auto offset = pBeamParams->GetProfilePlacement()->GetTransformFrom()->TransformPoint(&offset_);

  auto pBaseLine = pBeamParams->GetBaseline();

  auto pSectionPlacement = pBeamParams->GetProfilePlacementOnBaseline(pBaseLine->MinParameter);

  auto pSectionPlacementTransform = pSectionPlacement->GetTransformFrom();

  Renga::Point3D point = { offset.X, offset.Y, 0 };
  auto offsetPoint = pSectionPlacementTransform->TransformPoint(&point);
  auto baselineStart = pBaseLine->GetBeginPoint();

  return 
  {
    offsetPoint.X - baselineStart.X,
    offsetPoint.Y - baselineStart.Y,
    offsetPoint.Z - baselineStart.Z,
  };
}

Renga::IPlacement3DPtr BeamParamsClickHandler::getLevelPlacement(Renga::IModelObjectCollectionPtr pModelObjects, Renga::IModelObjectPtr pModelObject)
{
  Renga::ILevelObjectPtr pLevelObject;
  pModelObject->QueryInterface(&pLevelObject);

  auto parentLevelId = pLevelObject->LevelId;
  auto pParentLevel = pModelObjects->GetById(parentLevelId);

  Renga::ILevelPtr pLevel;
  pParentLevel->QueryInterface(&pLevel);

  return pLevel->Placement;
}

std::wstring BeamParamsClickHandler::buildBeamParamsDescription(Renga::IBeamParamsPtr pBeamParams, Renga::IBeamStyleManagerPtr pBeamStyleManager, Renga::IPlacement3DPtr pLevelPlacement)
{
  std::wstring result_output = L"  Beam Param:\n";

  auto pTransformIntoGlobal = pLevelPlacement->GetTransformFrom();
  auto pTransformIntoGlobalWithVerticalOffset = pTransformIntoGlobal->GetCopy();
  Renga::Vector3D vec = { 0, 0, pBeamParams->VerticalOffset };
  pTransformIntoGlobalWithVerticalOffset->AddTranslation(&vec);

  // getting the baseline of the beam
  auto pBaseLineInLevelSpace = pBeamParams->GetBaseline();

  // getting beam profile
  auto pProfile = pBeamStyleManager->GetBeamStyle(pBeamParams->StyleId)->Profile;
  auto pProfilePlacement = pBeamParams->GetProfilePlacement();

  // curves in the global coordinate system

  // getting the baseline of the beam in the global coordinate system
  auto pBaseLine = pBaseLineInLevelSpace->GetTransformed(pTransformIntoGlobal);

  // getting the center line of the beam in the global coordinate system
  auto vecOffset = GetBaselineOffset(pBeamParams, { 0, 0 });
  auto pCenterLine = pBaseLineInLevelSpace->GetOffseted(&vecOffset)->GetTransformed(pTransformIntoGlobalWithVerticalOffset);

  // getting the center line of mass of the beam in the global coordinate system
  vecOffset = GetBaselineOffset(pBeamParams, pProfile->GetCenterOfMass());
  auto pCenterOfMassPath = pBaseLineInLevelSpace->GetOffseted(&vecOffset)->GetTransformed(pTransformIntoGlobalWithVerticalOffset);

  // getting the length of the beam
  result_output += L"  beam length = " + std::to_wstring(pCenterOfMassPath->GetLength()) + L"\n\n";

  // getting the parameter of the corresponding beam start
  double parameter = pBaseLineInLevelSpace->MinParameter;

  auto pProfilePlacementOnBaseline = pBeamParams->GetProfilePlacementOnBaseline(parameter);
  for (int i = 0; i < pProfile->Regions->Count; ++i)
  {
    auto pCurrentRegion = pProfile->Regions->Get(i);
    for (int j = 0; j < pCurrentRegion->GetContourCount(); ++j)
    {
      auto currentContour = pCurrentRegion->GetContour(j)->GetTransformed(pProfilePlacement->GetTransformFrom());

      // obtaining the curve of one of the regions of the beam profile at the beginning of the beam
      auto pSectionCurve = currentContour->CreateCurve3D(pProfilePlacementOnBaseline)->GetTransformed(pTransformIntoGlobalWithVerticalOffset);
    }
  }
  return result_output;
}

std::wstring BeamParamsClickHandler::addPropertyToLevelObjects()
{
  auto pProject = m_pApplication->GetProject();

  if (!pProject)
    return L"The Renga project is null";

  std::wstring result_output = L"Beam Params:\n";

  auto pBeamStyleManager = pProject->BeamStyleManager;

  auto pModelObjects = pProject->Model->GetObjects();
  for (int i = 0; i < pModelObjects->Count; i++)
  {
    auto pModelObject = pModelObjects->GetByIndex(i);

    Renga::IBeamParamsPtr pBeamParams;
    pModelObject->QueryInterface(&pBeamParams);
    if (pBeamParams)
    {
      auto pLevelPlacement = getLevelPlacement(pModelObjects, pModelObject);
      result_output += buildBeamParamsDescription(pBeamParams, pBeamStyleManager, pLevelPlacement);
    }
  }
  return result_output;
}
